src/merkle_tree/mod.rs

```diff
@@ -1,253 +1,403 @@
+/// For this release, padding will not be supported because of security concerns: if the leaf hash and two to one hash uses same underlying
+/// CRH, a malicious prover can prove a leaf while the actual node is an inner node. In the future, we can prefix leaf hashes in different layers to
+/// solve the problem.
 pub struct MerkleTree<P: Config> {
-    tree: Vec<Digest<P>>,
-    padding_tree: Vec<(Digest<P>, Digest<P>)>,
-    parameters: <P::H as FixedLengthCRH>::Parameters,
-    root: Option<Digest<P>>,
+    /// stores the non-leaf nodes in level order. The first element is the root node.
+    non_leaf_nodes: Vec<TwoToOneDigest<P>>,
+    /// store the hash of leaf nodes from left to right
+    leaf_nodes: Vec<LeafDigest<P>>,
+    /// Store the two-to-one hash parameters
+    two_to_one_hash_param: TwoToOneParam<P>,
+    /// Store the leaf hash parameters
+    leaf_hash_param: LeafParam<P>,
+    /// Stores the height of the MerkleTree
+    height: usize,
 }
 
 impl<P: Config> MerkleTree<P> {
-    pub const HEIGHT: u8 = P::HEIGHT as u8;
-
-    pub fn blank(parameters: Parameters<P>) -> Self {
-        MerkleTree {
-            tree: Vec::new(),
-            padding_tree: Vec::new(),
-            root: None,
-            parameters,
-        }
+    /// Create an empty merkle tree such that all leaves are zero-filled.
+    pub fn blank(
+        leaf_hash_param: &LeafParam<P>,
+        two_to_one_hash_param: &TwoToOneParam<P>,
+        height: usize,
+    ) -> Result<Self, crate::Error> {
+        let leaf = vec![0u8; P::LeafHash::INPUT_SIZE_BITS / 8];
+        let leaves = vec![leaf; 1 << (height - 1)];
+        Self::new(leaf_hash_param, two_to_one_hash_param, &leaves)
     }
 
-    pub fn new<L: ToBytes>(parameters: Parameters<P>, leaves: &[L]) -> Result<Self, crate::Error> {
-        let new_time = start_timer!(|| "MerkleTree::New");
-
-        let last_level_size = leaves.len().next_power_of_two();
-        let tree_size = 2 * last_level_size - 1;
-        let tree_height = tree_height(tree_size);
-        assert!(tree_height as u8 <= Self::HEIGHT);
-
-        // Initialize the merkle tree.
-        let mut tree = Vec::with_capacity(tree_size);
-        let empty_hash = hash_empty::<P::H>(&parameters)?;
-        for _ in 0..tree_size {
-            tree.push(empty_hash.clone());
-        }
-
-        // Compute the starting indices for each level of the tree.
+    /// Returns a new merkle tree. `leaves.len()` should be power of two.
+    pub fn new<L: ToBytes>(
+        leaf_hash_param: &LeafParam<P>,
+        two_to_one_hash_param: &TwoToOneParam<P>,
+        leaves: &[L],
+    ) -> Result<Self, crate::Error> {
+        let leaf_nodes_size = leaves.len(); // size of the leaf layer
+        assert!(
+            leaf_nodes_size.is_power_of_two(),
+            "`leaves.len() should be power of two"
+        );
+        let non_leaf_nodes_size = leaf_nodes_size - 1;
+
+        let tree_height = tree_height(non_leaf_nodes_size + leaf_nodes_size);
```

Comment:
If the number of leaves is greater than `2 ** 16` this would overflow (in the worst-case scenario), limiting us to trees with a height of `16`. It would be great if we could use the full capacity of the `Vec`, so maybe we could convert them to `u64`, add them and then back to `usize` since `tree_height` won't be greater than `32`

Reply:
Same as below, I was confused a tiny bit. It's `2**31`. The point I was trying to make is that we are not using the full capacity of the `Vec`, but I'm fine with it if you are